Add pTypeManaged/sTypeManagedCounter devices for plugins and JSON

guillaumezin opened the pull request:
Mainly for counter externally managed (Linky, Gazpar) where only history can be grabbed and that will be managed by scripts

This is another implementation proposal following discussion in PR #2271

guillaumezin committed:
Add pTypeManaged/sTypeManagedCounter devices for plugins and JSON
Mainly for counter externally managed (Linky, Gazpar) where only history can be grabed and that will be managed by scripts

gizmocuz commented:
Have not looked at the patch yet 100%, but please remove the new pTypeManaged thing and use the
pTypeGeneral with a new subtype (pTypeManaged)
There are to many pType's already in the system

gizmocuz commented on main/EventSystem.cpp:
```diff
@@ -1061,6 +1061,13 @@ void CEventSystem::GetCurrentMeasurementStates()
+		case pTypeManaged:
+			if (sitem.subType == sTypeManagedCounter)
+			{
+				utilityval = static_cast<float>(atof(sitem.sValue.c_str()));
```
This counter only has a counter and no actual usage ? Also not in the feature ?
I say thinking about the feature, there might be someone willing to use this and also report the current usage
So a sValue like COUNTER(number);USAGE(Watt) would be a good idea?

guillaumezin committed:
Move sTypeManagedCounter into pTypeGeneral
sTypeManagedCounter now handles sValue as Counter;Usage or Counter or -1.0;Usage

guillaumezin committed:
Change main type number of managed counter for dummy devices

guillaumezin committed:
Change subtype number of managed counter for dummy devices

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -3587,6 +3587,9 @@ uint64_t CSQLHelper::CreateDevice(const int HardwareID, const int SensorType, co
 		case sTypeCounterIncremental:		//Counter Incremental
 			DeviceRowIdx = UpdateValue(HardwareID, ID, 1, SensorType, SensorSubType, 12, 255, 0, "0", devname);
 			break;
+		case sTypeManagedCounter:		//Counter Incremental
```
I think the comment here is wrong, could be removed (also for the counter incremental if you would)

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -5176,6 +5194,100 @@ void CSQLHelper::UpdateUVLog()
+	std::string devname = sd[1];
+	_eSwitchType switchtype=(_eSwitchType) atoi(sd[2].c_str());
+
+	if (dayCalendar)
```
this can not be the right boolean for the Meter table, the meter table is for short-log, and tables ending with _Calendar are for day values

guillaumezin replied:
The name of the boolean is misleading, I will change it

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -5176,6 +5194,100 @@ void CSQLHelper::UpdateUVLog()
 }
 
+bool CSQLHelper::UpdateCalendarMeter(const int HardwareID, const char* DeviceID, const unsigned char unit, const unsigned char devType, const unsigned char subType, bool dayCalendar, float counter, float usage, const char* date)
+{
```
Maybe a small check if the 'date' field is correct, for instance it's length + dashes on the right possition

gizmocuz reviewed (commented):
Some small comments

guillaumezin committed:
Check date time format

gizmocuz commented:
Thanks!

gizmocuz commented:
Made some small changes... i will rebuild our old code as well to use longlong as the database fields are not floats

https://github.com/domoticz/domoticz/commit/043e46ea7375381d357570d5d690c7974c9a030a
